Heartbeat timeout after update 4.8.1 -> 4.11.1

After updating the package on the client side to version 4.11.1, I get all the time a heartbeat timeout. It happens in version 4.10.0, too.

![image](https://github.com/user-attachments/assets/39ad1d4d-b3ed-4dd4-b2c2-ff16cb0927d8)

Here my heartbeat config which is working fine in 4.8.1:

```js
    heartbeat: {
        message: 'ping',
        returnMessage: 'pong',
        timeout: 30000,
        interval: 20000,
      },
````

On the server side I am running python websockets in version 13.1:

The heartbeat reply is solved as follows:
```python
if message == 'ping':
  await websocket.send('pong')
  continue
```

Downgrading to 4.8.1 removes the issue.

Edit: A short investigation with wireshark showed that in version 4.11.1 the ping message is not send to the backend. Therefore it timeouts. In version 4.8.1 the incoming (from backend side seen) ping text message is shown and the reply from the backend with text message "pong" happens.

On question: My understanding of the source code is, that the timer for the heartbeat is set on `webSocketInstance.onopen` and defined in `heartbeat.ts`. Therefore the first heartbeat ping is sent after the defined intervall in the heartbeat-options? I ask this question because when the websocket connection is established to my backend tokens were exchanged to check if the connection is authenticated. If the first message the client sends an heartbeat-ping this message will be dropped.

I appreciate our effort and I like to test your changes. After installing the package with the changes direct from git my react project is not able to resolve the module longer. It is missing it in the src folder of the project: `Module not found: Error: Can't resolve 'react-use-websocket' in '/home/username/appname/src'`. It is definitly offtopic but after some hours of googleing and try & error I am not able to motivate it to work. Any ideas/hints?

The entry in the `package.json` of the project looks like `"react-use-websocket": "github:w666/react-use-websocket"`.

--- a/src/lib/attach-shared-listeners.ts
+++ b/src/lib/attach-shared-listeners.ts
@@ -46,7 +46,8 @@ const bindOpenHandler = (
   heartbeatOptions?: boolean | HeartbeatOptions
 ) => {
   webSocketInstance.onopen = (event: WebSocketEventMap['open']) => {
-    getSubscribers(url).forEach(subscriber => {
+    const subscribers = getSubscribers(url);
+    subscribers.forEach(subscriber => {
       subscriber.reconnectCount.current = 0;
       if (subscriber.optionsRef.current.onOpen) {
         subscriber.optionsRef.current.onOpen(event);
@@ -58,9 +59,11 @@ const bindOpenHandler = (
 
       if (heartbeatOptions && webSocketInstance instanceof WebSocket) {
         subscriber.lastMessageTime.current = Date.now();
-        heartbeat(webSocketInstance, subscriber.lastMessageTime, typeof heartbeatOptions === 'boolean' ? undefined : heartbeatOptions,);
       }
     });
+    if (heartbeatOptions && webSocketInstance instanceof WebSocket) {
+      heartbeat(webSocketInstance, subscribers.map(subscriber => subscriber.lastMessageTime), typeof heartbeatOptions === 'boolean' ? undefined : heartbeatOptions,);
+    }
   };
 };
 

--- a/src/lib/heartbeat.ts
+++ b/src/lib/heartbeat.ts
@@ -2,25 +2,42 @@ import { MutableRefObject } from "react";
 import { DEFAULT_HEARTBEAT } from "./constants";
 import { HeartbeatOptions } from "./types";
 
-export function heartbeat(ws: WebSocket, lastMessageTime: MutableRefObject<number>, options?: HeartbeatOptions): () => void {
+function getLastMessageTime(lastMessageTime: MutableRefObject<number> | MutableRefObject<number>[]): number {
+  if (Array.isArray(lastMessageTime)) {
+    return lastMessageTime.reduce((p, c) => { return (p.current > c.current) ? p : c; }).current;
+  }
+  return lastMessageTime.current
+}
+
+export function heartbeat(ws: WebSocket, lastMessageTime: MutableRefObject<number> | MutableRefObject<number>[], options?: HeartbeatOptions): () => void {
   const {
     interval = DEFAULT_HEARTBEAT.interval,
     timeout = DEFAULT_HEARTBEAT.timeout,
     message = DEFAULT_HEARTBEAT.message,
   } = options || {};
 
+  // how often check interval between ping messages
+  // minimum is 100ms
+  // maximum is ${interval / 10}ms
+  const intervalCheck = Math.max(100, interval / 10);
+
+  let lastPingSentAt = Date.now();
+
   const heartbeatInterval = setInterval(() => {
-    if (lastMessageTime.current + timeout <= Date.now()) {
-      console.warn(`Heartbeat timed out, closing connection, last message was seen ${Date.now() - lastMessageTime.current}ms ago`);
+    const timeNow = Date.now();
+    const lastMessageReceivedAt = getLastMessageTime(lastMessageTime);
+    if (lastMessageReceivedAt + timeout <= timeNow) {
+      console.warn(`Heartbeat timed out, closing connection, last message received ${timeNow - lastMessageReceivedAt}ms ago, last ping sent ${timeNow - lastPingSentAt}ms ago`);
       ws.close();
     } else {
-      if (lastMessageTime.current + interval <= Date.now()) {
+      if (lastMessageReceivedAt + interval <= timeNow && lastPingSentAt + interval <= timeNow) {
         try {
           if (typeof message === 'function') {
             ws.send(message());
           } else {
             ws.send(message);
           }
+          lastPingSentAt = timeNow;
         } catch (err: unknown) {
           console.error(`Heartbeat failed, closing connection`, err instanceof Error ? err.message : err);
           ws.close();
@@ -28,12 +45,11 @@ export function heartbeat(ws: WebSocket, lastMessageTime: MutableRefObject<numbe
 
       }
     }
-  }, interval);
+  }, intervalCheck);
 
   ws.addEventListener("close", () => {
     clearInterval(heartbeatInterval);
   });
-
 
   return () => { };
 }

--- a/src/lib/use-websocket.test.ts
+++ b/src/lib/use-websocket.test.ts
@@ -614,7 +614,7 @@ test.each([false, true])('Options#heartbeat, if provided, close websocket if no 
   options.heartbeat = {
     message: 'ping',
     timeout: 1000,
-    interval: 500,
+    interval: 400,
   };
   options.share = shareOption;
 
@@ -627,6 +627,7 @@ test.each([false, true])('Options#heartbeat, if provided, close websocket if no 
   }
   await server.connected;
   await sleep(1600);
+  expect(server.messages).toEqual(['ping', 'ping'])
   expect(result.current.readyState).toBe(WebSocket.CLOSED);
 });
 
@@ -681,5 +682,41 @@ test.each([false, true])('Options#heartbeat, if provided, lastMessage is updated
   server.send('ping');
   expect(result.current.lastMessage?.data).toBe('ping');
 });
+
+test.each([false, true])('Options#heartbeat, can handle case when interval is very close to timeout', async (shareOption) => {
+  options.heartbeat = {
+    message: "ping",
+    returnMessage: "pong",
+    timeout: 1000,
+    interval: 800,
+  };
+  options.share = shareOption;
+
+  const { result } = renderHook(() => useWebSocket(URL, options));
+
+  if (shareOption) {
+    renderHook(() => useWebSocket(URL, options));
+  }
+
+  await server.connected;
+  await sleep(50);
+  expect(result.current.readyState).toEqual(ReadyState.OPEN);
+
+  result.current.sendMessage("token");
+  await sleep(50);
+  expect(server).toHaveReceivedMessages(["token"]);
+
+  server.send("authorized");
+  await sleep(50);
+  expect(result.current.lastMessage?.data).toEqual("authorized");
+
+  await sleep(850);
+  // it does not matter is it shared mode or not, in both cases only one timer is used that means only one ping message is sent for all subscribers
+  expect(server.messages).toEqual(["token", "ping"]);
+  server.send("pong");
+
+  expect(result.current.readyState).toBe(WebSocket.OPEN);
+}
+);
 
 // //TODO: Write companion tests for useSocketIO

